File: core/pathfinding.py
import heapq

def heuristic(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

def reconstruct_path(came_from, current):
    path = [current]
    while current in came_from:
        current = came_from[current]
        path.append(current)
    path.reverse()
    return path

def a_star(grid_world, start, goal):
    width, height = grid_world.width, grid_world.height
    open_set = []
    heapq.heappush(open_set, (0, start))

    came_from = {}
    g_score = {start: 0}
    f_score = {start: heuristic(start, goal)}

    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

    while open_set:
        _, current = heapq.heappop(open_set)

        if current == goal:
            return reconstruct_path(came_from, current)
        
        for dx, dy in directions:
            neighbour = (current[0] + dx, current[1] + dy)

            if not grid_world.is_within_bounds(*neighbour):
                continue

            tentative_g_score = g_score[current] + 1

            if neighbour not in g_score or tentative_g_score < g_score[neighbour]:
                came_from[neighbour] = current
                g_score[neighbour] = tentative_g_score
                f_score[neighbour] = tentative_g_score + heuristic(neighbour, goal)
                heapq.heappush(open_set, (f_score[neighbour], neighbour))

    return None

File: C_SHARP/Pathfinding.cs
using OpenNPC_CSharp_ver;
using System;
using System.Collections.Generic;
using System.Linq;

public static class Pathfinding
{
	public static int Heuristic((int x, int y) a, (int x, int y) b)
	{
		return Math.Abs(a.x - b.x) + Math.Abs(a.y - b.y);
	}

	public static List<(int x, int y)> AStar(GridWorld world, (int x, int y) start, (int x, int y) goal)
	{
		var openSet = new List<(int x, int y)> { start };
		var cameFrom = new Dictionary<(int, int), (int, int)>();
		var gScore = new Dictionary<(int, int), int> { [start] = 0 };
		var fScore = new Dictionary<(int, int), int> { [start] = Heuristic(start, goal) };

		while (openSet.Count > 0)
		{
			var current = openSet.OrderBy(n => fScore.ContainsKey(n) ? fScore[n] : int.MaxValue).First();
			
			if (current == goal)
				return ReconstructPath(cameFrom, current);

			openSet.Remove(current);

			foreach (var neighbor in GetNeighbors(world, current))
			{
				int tentativeG = gScore[current] + 1;

				if (!gScore.ContainsKey(neighbor) || tentativeG < gScore[neighbor])
				{
					cameFrom[neighbor] = current;
					gScore[neighbor] = tentativeG;
					fScore[neighbor] = tentativeG + Heuristic(neighbor, goal);

					if (!openSet.Contains(neighbor))
						openSet.Add(neighbor);
				}
			}
		}
		return null;
	}

	private static List<(int x, int y)> GetNeighbors(GridWorld world, (int x, int y) node)
	{
		var dirs = new (int dx, int dy)[] { (1, 0), (-1, 0), (0, 1), (0, -1) };
		var neighbors = new List<(int x, int y)>();

		foreach (var (dx, dy) in dirs)
		{
			int nx = node.x + dx;
			int ny = node.y + dy;
			if (world.IsWalkable(nx, ny))
				neighbors.Add((nx, ny));
		}
		return neighbors;
	}

	private static List<(int x, int y)> ReconstructPath(Dictionary<(int, int), (int, int)> cameFrom, (int x, int y) current)
	{
		var totalPath = new List<(int, int)> { current };
		while (cameFrom.ContainsKey(current))
		{
			current = cameFrom[current];
			totalPath.Insert(0, current);
		}
		return totalPath;
	}
}